### PR TITLE
Closes #2265 Add manual migration lookup process plugin for migrate.

### LINKED ISSUE #2265: Add migrate process plugin to look up manually migrated (Copy Pasted) content.
During a recent migration, I had the need to reference manually migrated entities within a programmatic migration.

## Motivation
Possibly contribute the process plugin that I made to Quickstart so that others can benefit.

### PR DESCRIPTION
**Pull request by trackleft**
## Description

See #2265 

Add a migrate process plugin that allows looking up data in the source database that can be used in conjunction with something like  `entity_lookup` or `entity_generate` process plugins for content that is not being tracked in a migrate map database table (Migrated using the migrate module).

## Related issues
Closes #2265 

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [x] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: modules/custom/az_core/src/Plugin/migrate/process/ManualMigrationLookup.php [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+      switch ($source_entity_type) {
+        case 'node':
+          // Lookup of content type.
+          $value = Database::getConnection('default', 'migrate')
+            ->query('SELECT title FROM {node} WHERE nid = :nid', [':nid' => $id])
+            ->fetchField();
+          break;
+
+        case 'taxonomy_term':
+            // Lookup of content type.
```
Comment (trackleft) [inline modules/custom/az_core/src/Plugin/migrate/process/ManualMigrationLookup.php:98]:
```suggestion
            // Lookup of taxonomy term.
```

File: modules/custom/az_core/src/Plugin/migrate/process/ManualMigrationLookup.php [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+ *
+ * @code
+ *  process:
+ *   field_tags:
+ *    plugin: sub_process
+ *    source: field_tags
+ *    process:
+ *      delta: delta
+ *      target_id:
+ *        - plugin: az_manual_migration_lookup
```
Comment (trackleft) [inline modules/custom/az_core/src/Plugin/migrate/process/ManualMigrationLookup.php:57]: could just call this `source_db_table_value_lookup` and then add fields and table name to the config.

File: modules/custom/az_core/src/Plugin/migrate/process/ManualMigrationLookup.php [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+use Drupal\migrate\MigrateExecutableInterface;
+use Drupal\migrate\MigrateSkipProcessException;
+use Drupal\migrate\ProcessPluginBase;
+use Drupal\migrate\Row;
+
+/**
+ * Process Plugin to handle content that was manually migrated.
+ *
+ * @MigrateProcessPlugin(
+ *   id = "az_manual_migration_lookup"
```
Comment (trackleft) [inline modules/custom/az_core/src/Plugin/migrate/process/ManualMigrationLookup.php:15]:
```suggestion
 *   id = "az_manual_migration_lookup"
```

could just call this `source_db_table_value_lookup` and then add fields and table name to the config.

Comment (trackleft) [reply modules/custom/az_core/src/Plugin/migrate/process/ManualMigrationLookup.php:15]: would look like this in practice 

```
          - plugin: source_db_table_value_lookup_by_id
            select_fields: title
            select_table: node
            select_index: nid
            source: target_id
 ```

Comment (trackleft) [reply modules/custom/az_core/src/Plugin/migrate/process/ManualMigrationLookup.php:15]: Probably too general and complicated to be useful?

File: modules/custom/az_core/src/Plugin/migrate/process/ManualMigrationLookup.php [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+ * )
+ * Only works when the source database is a Drupal 7 database.
+ *
+ * This plugin looks up content from the source database and returns the field
+ * data for the found entity. This requires an entity id from the source site.
+ * This is useful for content that was manually migrated (Copy and Pasted) and not
+ * tied to a migration.
```
Comment (joeparsons) [inline modules/custom/az_core/src/Plugin/migrate/process/ManualMigrationLookup.php:22]: PHPCS warning: 
```
FILE: ...tom/az_core/src/Plugin/migrate/process/ManualMigrationLookup.php
----------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
----------------------------------------------------------------------
 21 | WARNING | Line exceeds 80 characters; contains 82
    |         | characters (Drupal.Files.LineLength.TooLong)
----------------------------------------------------------------------
```

```suggestion
 * This is useful for content that was manually migrated (Copy and Pasted) and
 * not tied to a migration.
```

File: modules/custom/az_core/src/Plugin/migrate/process/ManualMigrationLookup.php [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+ *
+ * @return string source entity identifier Example: node title.
```
Comment (joeparsons) [inline modules/custom/az_core/src/Plugin/migrate/process/ManualMigrationLookup.php:24]: Nit-picky suggestion:
```suggestion
 * @return string source entity label (Example: node title).
```

File: modules/custom/az_core/src/Plugin/migrate/process/ManualMigrationLookup.php [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+namespace Drupal\az_core\Plugin\migrate\process;
+
+use Drupal\Core\Database\Database;
+use Drupal\migrate\MigrateExecutableInterface;
+use Drupal\migrate\MigrateSkipProcessException;
+use Drupal\migrate\ProcessPluginBase;
+use Drupal\migrate\Row;
+
+/**
+ * Process Plugin to handle content that was manually migrated.
```
Comment (joeparsons) [inline modules/custom/az_core/src/Plugin/migrate/process/ManualMigrationLookup.php:12]: Nit-pick:
```suggestion
 * Process plugin to handle content that was manually migrated.
```

**Review by joeparsons — APPROVED** [review]

**Review by tadean — APPROVED** [review]

**Pull request merged**